packages/core/src/renderer/Renderer.ts

```diff
@@ -155,23 +201,83 @@ export const RenderInteractive = async (
   const onDrag = (id: string, dx: number, dy: number) => {
     updateState(dragUpdate(state, id, dx, dy));
   };
-  for (const shape of state.computeShapes(state.varyingValues)) {
-    svg.appendChild(
-      await DraggableShape(
-        {
-          shape,
-          labels: state.labelCache,
-          canvasSize: state.canvas.size,
-          pathResolver,
-        },
-        onDrag,
-        svg
-      )
-    );
-  }
+  const shapes = state.computeShapes(state.varyingValues);
+  const graph = makeGroupGraph(shapes);
```

Comment:
The `Render*` functions are called quite frequently in clients, so I don't think it's great to re-compute the group graph every single time, especially when they largely stay the same. One alternative is to cache a static group graph with shape names as nodes in `State` and simply index into `shapes` at render time.

Reply:
Done!